codice/ddf
# DDF-3997 Create a service to support an incoming query and outbound format for generating files

#### What does this PR do?
- Refactors much of the existing QueryApplication to support handler classes, allowing for unit testing on Spark endpoints.
- Creates a `/internal/cql/transform/{transformerId}` endpoint used to export a file from a result set.
- The result set from the query response will be transformed to the matching id, and returned in the http response in the file format that matches the transformer's mime-type.
- Optional arguments can be passed to the transformer from the http request body using an `arguments` json key.

#### Who is reviewing it? 

@brianfelix 
@leo-sakh 
@bellcc
@jrnorth 
#### Select relevant component teams: 
https://github.com/orgs/codice/teams
#### Ask 2 committers to review/merge the PR and tag them here.

@millerw8 
@bdeining 
#### How should this be tested?


- (Optional) Disable web sockets in Admin > System > Catalog UI Search
- Populate local catalog with metacard(s) (via postman or ui uploader)
- Search using a transformer id (I recommend `/cql/transform/xml`). You can do this in postman, however I was able to do this easier with the `Requestly` chrome extension, where you replace the host in a request from `cql` to `cql/transform/xml`, and execute that query from the run button in DDF ui (and view the response data in network dev tools).
- Example request body: (src and cql required keys)
```
{
    "src":"cache",
    "start":1,
    "count":250,
    "cql":"((anyText ILIKE '%') AND ((\"metacard_source\" = 'ddf.distribution')))",
    "sorts": [
        {
            "attribute":"modified",
            "direction":"descending"
            
        }
    ],
    "id":"6d9ed55a-1b6c-4b94-992f-c4e7374ae3dc"
}
```
- In network dev tools or postman ensure the response header returns `attachment;filename=export-{time}{fileformat}` (for ex: `attachment;filename=export-2018-07-30T20:36:15.302Z.xml`) and contains the metacard(s) data in proper transformation data format (in my case xml).

#### Any background context you want to provide?

Future tickets would like a result set to be exported to various file formats. Transformers will need to be made for these formats, and this service will allow the UI to perform this operation to access those transformers.

#### What are the relevant tickets?
[DDF-3997](https://codice.atlassian.net/browse/DDF-3997)
#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
